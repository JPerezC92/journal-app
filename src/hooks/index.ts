export * from "./useForm";
export * from "./useAuthentication";
