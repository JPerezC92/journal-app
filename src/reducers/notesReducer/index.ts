export * from "./notesReducer";
export * from "./notesThunks";
