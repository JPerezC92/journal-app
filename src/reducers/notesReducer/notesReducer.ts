import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface Note {
  id: string;
  title: string;
  body: string;
  imageUrl: string | null;
  date: number;
}

interface NotesState {
  notes: Note[];
  active: Note | null;
}
const initialState: NotesState = { notes: [], active: null };

const notesSlice = createSlice({
  name: "[Notes]",
  initialState,
  reducers: {
    addNewNote: (state, { payload }: PayloadAction<Note>) => {
      state.notes = [...state.notes, payload];
    },
    setNoteActive: (state, action: PayloadAction<Partial<Note>>) => {
      state.active = { ...state.active!, ...action.payload };
    },
    setNotes: (state, action: PayloadAction<Note[]>) => {
      state.notes = action.payload;
    },
    refreshNote: (state, { payload }: PayloadAction<Note>) => {
      state.notes = state.notes.map((note) =>
        note.id === payload.id ? payload : note
      );
    },

    deleteNote: (state, { payload }: PayloadAction<string>) => {
      state.notes = state.notes.filter((note) => note.id !== payload);
      state.active = null;
    },
    notesLogoutCleaning: (state) => {
      state.notes = [];
      state.active = null;
    },
  },
});

export const notesActions = notesSlice.actions;

export const notesReducer = notesSlice.reducer;
