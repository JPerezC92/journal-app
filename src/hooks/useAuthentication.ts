import { useDispatch } from "react-redux";
import { startLogin, startLogout } from "../reducers";
import { AuthService } from "../services";

interface UseAuthentication {
  (): {
    handleLogout: () => void;
    handleLoginWithEmailAndPassword: (email: string, password: string) => void;
    handleGoogleLogin: () => void;
  };
}

export const useAuthentication: UseAuthentication = () => {
  const dispatch = useDispatch();

  const handleLoginWithEmailAndPassword = (email: string, password: string) => {
    dispatch(
      startLogin(() => AuthService.loginWithEmailAndPassword(email, password))
    );
  };

  const handleGoogleLogin = () => {
    dispatch(startLogin(AuthService.loginWithGoogle));
  };

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return { handleLogout, handleGoogleLogin, handleLoginWithEmailAndPassword };
};
