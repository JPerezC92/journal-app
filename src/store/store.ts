import { configureStore } from "@reduxjs/toolkit";
import { uiReducer, authReducer, notesReducer } from "../reducers";

export const reducers = {
  authReducer,
  uiReducer,
  notesReducer,
};

export const store = configureStore({
  reducer: reducers,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
