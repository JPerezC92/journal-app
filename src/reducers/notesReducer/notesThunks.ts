import { createAsyncThunk } from "@reduxjs/toolkit";
import { db } from "../../firebase";
import { RootState } from "../../store/store";
import { Note, notesActions } from "./notesReducer";

enum NotesThunks {
  START_NEW_NOTE = "START_NEW_NOTE",
}

export const startNewNote = createAsyncThunk(
  NotesThunks.START_NEW_NOTE,
  async (_, { dispatch, getState }) => {
    const { uid } = (getState() as RootState).authReducer.user;
    const newNote: Note = {
      id: null,
      title: "",
      body: "",
      date: new Date().getTime(),
      imageUrl: null,
    };

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(notesActions.setNoteActive({ id: docRef.id, note: newNote }));
  }
);
