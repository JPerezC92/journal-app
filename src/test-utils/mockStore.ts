import {
  AnyAction,
  CombinedState,
  combineReducers,
  configureStore,
  createSlice,
} from "@reduxjs/toolkit";
import { reducers } from "../store/store";

const storeCleanSlice = createSlice({
  name: "store",
  initialState: {},
  reducers: {
    clean: () => {},
  },
});

export const storeCleanActions = storeCleanSlice.actions;
const storeCleanReducer = storeCleanSlice.reducer;
const combinedReducers = combineReducers({ ...reducers, storeCleanReducer });

const testReducers = (
  state: CombinedState<any> | undefined,
  action: AnyAction
) => {
  if (action.type === "store/clean") {
    state = undefined;
  }
  return combinedReducers(state, action);
};

export const mockStore = configureStore({
  reducer: testReducers,
});

export type MockRootState = ReturnType<typeof mockStore.getState>;
export type MockAppDispatch = typeof mockStore.dispatch;
