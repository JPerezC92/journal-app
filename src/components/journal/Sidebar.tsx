import { useDispatch, useSelector } from "react-redux";
import { useAuthentication } from "../../hooks";
import { startNewNote } from "../../reducers/notesReducer/notesThunks";
import { RootState } from "../../store/store";
import JournalEntries from "./JournalEntries";

const Sidebar = () => {
  const dispatch = useDispatch();
  const auth = useSelector((state: RootState) => state.authReducer);
  const { displayName } = auth.user;

  const authenticationHook = useAuthentication();

  const handleAddNewEntry = () => {
    dispatch(startNewNote());
  };

  return (
    <aside className="journal__sidebar">
      <div className="journal__sidebar-navbar mt-1">
        <h3>
          <i className="far fa-moon"> </i>
          <span> {displayName}</span>
        </h3>

        <button
          type="button"
          className="btn"
          onClick={authenticationHook.handleLogout}
        >
          Logout
        </button>
      </div>
      <div
        aria-label="button"
        className="journal__new-entry"
        onClick={handleAddNewEntry}
      >
        <i className="far fa-calendar-plus fa-5x"></i>
        <p className="mt-5">New entry</p>
      </div>
      <JournalEntries />
    </aside>
  );
};

export default Sidebar;
