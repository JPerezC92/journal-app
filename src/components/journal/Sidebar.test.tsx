import renderer from "react-test-renderer";
import { act, cleanup, fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { store } from "../../store/store";
import { mockStore, storeCleanActions } from "../../test-utils/mockStore";
import * as notesThunks from "../../reducers/notesReducer/notesThunks";
import * as hooks from "../../hooks/useAuthentication";
import Sidebar from "./Sidebar";

describe("Test on <Sidebar />", () => {
  let dispatch = mockStore.dispatch;
  beforeEach(() => {
    dispatch = mockStore.dispatch;
  });

  afterEach(() => {
    dispatch(storeCleanActions.clean());
    jest.restoreAllMocks();
    cleanup();
  });

  test("should match the snapshot correctly", () => {
    let tree;
    renderer.act(() => {
      tree = renderer
        .create(
          <Provider store={store}>
            <MemoryRouter>
              <Sidebar />
            </MemoryRouter>
          </Provider>
        )
        .toJSON();
    });

    expect(tree).toMatchSnapshot();
  });

  test("should call startNewNote", async () => {
    jest.spyOn(notesThunks, "startNewNote");

    const component = render(
      <Provider store={mockStore}>
        <MemoryRouter>
          <Sidebar />
        </MemoryRouter>
      </Provider>
    );

    const newEntryButton = component.getByLabelText("button");

    act(() => {
      fireEvent.click(newEntryButton);
    });

    expect(notesThunks.startNewNote).toHaveBeenCalledTimes(1);
  });

  test("should calld handleLogout", async () => {
    const handleLogout = jest.fn();
    jest
      .spyOn(hooks, "useAuthentication")
      .mockImplementation(() => ({ handleLogout }));

    const component = render(
      <Provider store={mockStore}>
        <MemoryRouter>
          <Sidebar />
        </MemoryRouter>
      </Provider>
    );

    const logoutButton = component.getByRole("button");

    act(() => {
      fireEvent.click(logoutButton);
    });

    expect(handleLogout).toHaveBeenCalledTimes(1);
  });
});
