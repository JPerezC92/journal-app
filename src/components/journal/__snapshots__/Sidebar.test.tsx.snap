// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <Sidebar /> should match the snapshot correctly 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar mt-1"
  >
    <h3>
      <i
        className="far fa-moon"
      >
         
      </i>
      <span>
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      Logout
    </button>
  </div>
  <div
    aria-label="button"
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
